File: src/ClientModel/MethodGroupClient.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System.Collections.Generic;

namespace AutoRest.Java.Model
{
    /// <summary>
    /// The details of a group of methods within a ServiceClient.
    /// </summary>
    public class MethodGroupClient
    {
        /// <summary>
        /// Create a new MethodGroupClient with the provided properties.
        /// </summary>
        /// <param name="className">The name of the client's class.</param>
        /// <param name="interfaceName">The name of the client's interface.</param>
        /// <param name="implementedInterfaces">The interfaces that the client implements.</param>
        /// <param name="proxy">The REST API that the client will send requests to.</param>
        /// <param name="serviceClientName">The name of the ServiceClient that contains this MethodGroupClient.</param>
        /// <param name="variableType">The type of this MethodGroupClient when it is used as a variable.</param>
        /// <param name="variableName">The variable name for any instances of this MethodGroupClient.</param>
        /// <param name="clientMethods">The ClientMethods for this MethodGroupClient.</param>
        public MethodGroupClient(string package, string className, string interfaceName, IEnumerable<string> implementedInterfaces, Proxy proxy, string serviceClientName, string variableType, string variableName, IEnumerable<ClientMethod> clientMethods)
        {
            Package = package;
            ClassName = className;
            InterfaceName = interfaceName;
            ImplementedInterfaces = implementedInterfaces;
            Proxy = proxy;
            ServiceClientName = serviceClientName;
            VariableType = variableType;
            VariableName = variableName;
            ClientMethods = clientMethods;
        }

        public string Package { get; }

        /// <summary>
        /// Get the name of this client's class.
        /// </summary>
        public string ClassName { get; }

        /// <summary>
        /// Get the name of this client's interface.
        /// </summary>
        public string InterfaceName { get; }

        /// <summary>
        /// Get the interfaces that the client implements.
        /// </summary>
        public IEnumerable<string> ImplementedInterfaces { get; }

        /// <summary>
        /// Get the REST API that this client will send requests to.
        /// </summary>
        public Proxy Proxy { get; }

        /// <summary>
        /// Get the name of the ServiceClient that contains this MethodGroupClient.
        /// </summary>
        public string ServiceClientName { get; }

        /// <summary>
        /// Get the type of this MethodGroupClient when it is used as a variable.
        /// </summary>
        public string VariableType { get; }

        /// <summary>
        /// Get the variable name for any instances of this MethodGroupClient.
        /// </summary>
        public string VariableName { get; }

        /// <summary>
        /// The client method overloads for this MethodGroupClient.
        /// </summary>
        public IEnumerable<ClientMethod> ClientMethods { get; }

        /// <summary>
        /// Add this property's imports to the provided ISet of imports.
        /// </summary>
        /// <param name="imports">The set of imports to add to.</param>
        /// <param name="includeImplementationImports">Whether or not to include imports that are only necessary for method implementations.</param>
        public void AddImportsTo(ISet<string> imports, bool includeImplementationImports, JavaSettings settings)
        {
            if (!settings.IsFluent)
            {
                imports.Add($"{settings.Package}.{InterfaceName}");
            }

            if (includeImplementationImports)
            {
                ClassType proxyType = settings.IsAzureOrFluent ? ClassType.AzureProxy : ClassType.RestProxy;
                imports.Add(proxyType.FullName);
            }

            Proxy.AddImportsTo(imports, includeImplementationImports, settings);

            foreach (ClientMethod clientMethod in ClientMethods)
            {
                clientMethod.AddImportsTo(imports, includeImplementationImports, settings);
            }
        }
    }
}

File: mapper/src/main/java/com/azure/autorest/model/clientmodel/MethodGroupClient.java
package com.azure.autorest.model.clientmodel;

import com.azure.autorest.extension.base.plugin.JavaSettings;

import java.util.List;
import java.util.Set;

// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.


/** 
 The details of a group of methods within a ServiceClient.
*/
public class MethodGroupClient
{
    /**
     Create a new MethodGroupClient with the provided properties.

     @param package The package of the client's class.
     @param className The name of the client's class.
     @param interfaceName The name of the client's interface.
     @param implementedInterfaces The interfaces that the client implements.
     @param proxy The REST API that the client will send requests to.
     @param serviceClientName The name of the ServiceClient that contains this MethodGroupClient.
     @param variableType The type of this MethodGroupClient when it is used as a variable.
     @param variableName The variable name for any instances of this MethodGroupClient.
     @param clientMethods The ClientMethods for this MethodGroupClient.
    */
    public MethodGroupClient(String package_Keyword, String className, String interfaceName, List<String> implementedInterfaces, Proxy proxy, String serviceClientName, String variableType, String variableName, List<ClientMethod> clientMethods)
    {
        Package = package_Keyword;
        ClassName = className;
        InterfaceName = interfaceName;
        ImplementedInterfaces = implementedInterfaces;
        Proxy = proxy;
        ServiceClientName = serviceClientName;
        VariableType = variableType;
        VariableName = variableName;
        ClientMethods = clientMethods;
    }

    private String Package;
    public final String getPackage()
    {
        return Package;
    }

    /**
     Get the name of this client's class.
    */
    private String ClassName;
    public final String getClassName()
    {
        return ClassName;
    }

    /**
     Get the name of this client's interface.
    */
    private String InterfaceName;
    public final String getInterfaceName()
    {
        return InterfaceName;
    }

    /**
     Get the interfaces that the client implements.
    */
    private List<String> ImplementedInterfaces;
    public final List<String> getImplementedInterfaces()
    {
        return ImplementedInterfaces;
    }

    /**
     Get the REST API that this client will send requests to.
    */
    private Proxy Proxy;
    public final Proxy getProxy()
    {
        return Proxy;
    }

    /**
     Get the name of the ServiceClient that contains this MethodGroupClient.
    */
    private String ServiceClientName;
    public final String getServiceClientName()
    {
        return ServiceClientName;
    }

    /**
     Get the type of this MethodGroupClient when it is used as a variable.
    */
    private String VariableType;
    public final String getVariableType()
    {
        return VariableType;
    }

    /**
     Get the variable name for any instances of this MethodGroupClient.
    */
    private String VariableName;
    public final String getVariableName()
    {
        return VariableName;
    }

    /**
     The client method overloads for this MethodGroupClient.
    */
    private List<ClientMethod> ClientMethods;
    public final List<ClientMethod> getClientMethods()
    {
        return ClientMethods;
    }

    /**
     Add this property's imports to the provided ISet of imports.

     @param imports The set of imports to add to.
     @param includeImplementationImports Whether or not to include imports that are only necessary for method implementations.
    */
    public final void AddImportsTo(Set<String> imports, boolean includeImplementationImports, JavaSettings settings)
    {
        if (!settings.IsFluent && settings.GenerateClientInterfaces)
        {
            imports.add(String.format("%1$s.%2$s", settings.Package, getInterfaceName()));
        }

        if (includeImplementationImports)
        {
            ClassType proxyType = settings.IsAzureOrFluent ? ClassType.AzureProxy : ClassType.RestProxy;
            imports.add(proxyType.getFullName());
        }

        getProxy().AddImportsTo(imports, includeImplementationImports, settings);

        for (ClientMethod clientMethod : getClientMethods())
        {
            clientMethod.AddImportsTo(imports, includeImplementationImports, settings);
        }
    }
}